contracts/BridgeProxy.sol

```diff
@@ -0,0 +1,118 @@
+//SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./lib/Pausable.sol";
+import "./SharedStructs.sol";
+import "./lib/BoolTokenTransfer.sol";
+import "./access/AdminRole.sol";
+import "./access/BridgeRole.sol";
+
+contract BridgeProxy is Initializable, Pausable, BridgeRole {
+    using BoolTokenTransfer for IERC20;
+
+    uint256 public constant MIN_GAS_LIMIT_FOR_SC_CALL = 10_000_000;
+    uint256 public constant DEFAULT_GAS_LIMIT_FOR_REFUND_CALLBACK = 20_000_000;
+
+    mapping(uint256 => MvxTransaction) private pendingTransactions;
+    uint256 private lowestTxId;
+    uint256 private currentTxId;
+
+    function initialize() public initializer {
+        __BridgeRole_init();
+        __Pausable_init();
+    }
+
+    function __BridgeProxy__init_unchained() internal onlyInitializing {
+        lowestTxId = 0;
+    }
+
+    function deposit(MvxTransaction calldata txn) external payable whenNotPaused onlyBridge {
+        pendingTransactions[currentTxId] = txn;
+        currentTxId++;
+    }
+
+    function execute(uint256 txId) external whenNotPaused {
+        require(txId < currentTxId, "BridgeProxy: Invalid transaction ID");
+        MvxTransaction memory txn = pendingTransactions[txId];
+
+        require(txn.amount != 0, "BridgeProxy: No amount bridged");
+
+        if (txn.callData.length > 0) {
+            (bytes memory selector, uint256 gasLimit, bytes memory args) = abi.decode(
+                txn.callData,
+                (bytes, uint256, bytes)
+            );
+
+            if (selector.length == 0 || gasLimit == 0 || gasLimit < MIN_GAS_LIMIT_FOR_SC_CALL) {
+                _finishExecuteGracefully(txId, true);
+                return;
+            }
+
+            bytes memory data;
+            if (args.length > 0) {
+                data = abi.encodePacked(selector, args);
+            } else {
+                data = selector;
+            }
+
+            (bool success, ) = txn.recipient.call{ gas: gasLimit }(data);
```

Comment:
This is vulnerable to reentrancy, finish before calling, calling should be the last step. Also in this case, `_finish[...]` may not be the best name for this function

Reply:
actually only these lines should be moved above this contract call
```
 _updateLowestTxId();
delete pendingTransactions[txId];